src/wallet/memo.h

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2022 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_WALLET_MEMO_H
+#define ZCASH_WALLET_MEMO_H
+
+#include "util/strencodings.h"
+#include "zcash/Zcash.h"
+#include "tinyformat.h"
+
+
+enum class MemoError {
+    HexDecodeError,
+    MemoTooLong
+};
+
+typedef std::array<unsigned char, ZC_MEMO_SIZE> MemoBytes;
+
+class Memo {
+private:
+    MemoBytes value;
+public:
+    // initialize to default memo (no_memo), see section 5.5 of the protocol spec
+    Memo(): value({{0xF6}}) { }
+    Memo(MemoBytes value): value(value) {}
+
+    static Memo NoMemo() {
+        return Memo();
+    }
+
+    static std::variant<MemoError, Memo> FromHex(const std::string& memoHex) {
+        Memo result;
+        std::vector<unsigned char> rawMemo = ParseHex(memoHex.c_str());
+
+        // If ParseHex comes across a non-hex char, it will stop but still return results so far.
+        size_t slen = memoHex.length();
+        if (slen % 2 != 0 || (slen > 0 && rawMemo.size() != slen / 2)) {
```

Comment:
I think this is equivalent:
```suggestion
        if (slen != rawMemo.size() * 2) {
```

Reply:
Yeah, that’s certainly an improvement. I’m curious why the existing implementation wouldn’t error on negative even `slen` – seems like a bug. Definitely the kind of thing that should have been commented if not a bug!